src/types/list.rs

```diff
@@ -293,21 +295,45 @@ impl PyList {
     pub fn to_tuple(&self) -> &PyTuple {
         unsafe { self.py().from_owned_ptr(ffi::PyList_AsTuple(self.as_ptr())) }
     }
+
+    #[inline]
+    pub(crate) unsafe fn from_owned_ptr(py: Python<'py>, ptr: *mut ffi::PyObject) -> PyList<'py> {
+        PyList(PyAnyOwned::from_owned_ptr(py, ptr))
+    }
+
+    #[inline]
+    pub(crate) unsafe fn from_owned_ptr_or_err(
+        py: Python<'py>,
+        ptr: *mut ffi::PyObject,
+    ) -> PyResult<PyList<'py>> {
+        PyAnyOwned::from_owned_ptr_or_err(py, ptr).map(PyList)
+    }
+
+    fn into_non_null(self) -> NonNull<ffi::PyObject> {
+        self.0.into_non_null()
+    }
 }
 
-index_impls!(PyList, "list", PyList::len, PyList::get_slice);
+impl From<PyList<'_>> for Py<PyList<'static>> {
+    fn from(value: PyList<'_>) -> Self {
+        unsafe { Py::from_non_null(value.into_non_null()) }
+    }
+}
+
+// FIXME: can indexing be supported in an owned world?
+// index_impls!(PyList, "list", PyList::len, PyList::get_slice);
 
 /// Used by `PyList::iter()`.
-pub struct PyListIterator<'a> {
-    list: &'a PyList,
+pub struct PyListIterator<'py> {
```

Comment:
Since this is not a `#[pyclass]`, wouldn't the double indirection `&'a PyList<'py>` be closer to the usual Rust pattern for iterators borrowing the container? I think the iterator could still yield `PyAny<'py>` items though?